src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java

```diff
@@ -204,12 +215,102 @@ public InnerClassLambdaMetafactory(MethodHandles.Lookup caller,
     }
 
     private static String lambdaClassName(Class<?> targetClass) {
+        return createNameFromTargetClass(targetClass) + counter.incrementAndGet();
+    }
+
+    private static String createNameFromTargetClass(Class<?> targetClass) {
         String name = targetClass.getName();
         if (targetClass.isHidden()) {
             // use the original class name
             name = name.replace('/', '_');
         }
-        return name.replace('.', '/') + "$$Lambda$" + counter.incrementAndGet();
+        return name.replace('.', '/') + "$$Lambda$";
+    }
+
+    /**
+     * Calculate hash value of the given String in the same manner as the
+     * java.lang.StringUTF16#hashCode does with difference that hash value
+     * is long instead of int.
+     *
+     * @param name String for which method calculates hash value for
+     *
+     * @return a hash value for the given String
+     * */
+    private String fixedSizeStringHash(String name) {
+        long h = 0;
+        int length = name.length();
+        for (int i = 0; i < length; i++) {
+            h = 31 * h + name.charAt(i);
+        }
+
+        StringBuilder stable = new StringBuilder(Long.toString(Math.abs(h), Character.MAX_RADIX));
+        int stableLength = stable.length();
+
+        // We want all the hashed names to be of the same length. We pad some of them with the character 'a'
+        // to achieve this.
+        while (stableLength != targetLengthForStableName) {
+            stable.append(paddingCharacter);
+            stableLength++;
+        }
+
+        return stable.toString();
+    }
+
+    private String stableLambdaNameHash(String name) {
+        StringBuilder[] hashFragments = new StringBuilder[]{new StringBuilder(), new StringBuilder()};
+        int n = name.length();
+
+        for (int i = 0; i < n; i++) {
+            hashFragments[i % hashFragments.length].append(name.charAt(i));
+        }
+
+        StringBuilder stableNameHash = new StringBuilder();
+        for (StringBuilder sb : hashFragments) {
+            stableNameHash.append(fixedSizeStringHash(sb.toString()));
+        }
+
+        return stableNameHash.toString();
+    }
+
+    /**
+     * Creating stable name for lambda class.
+     * Parameters that are used to create stable name
+     * are a superset of the parameters that are used in
```

Comment:
I missed this bit the first two times I went over the PR :slightly_smiling_face: 
I've traced the call from `java.lang.invoke.LambdaProxyClassArchive#addToArchive` to the archiving in C:
```
void SystemDictionaryShared::add_lambda_proxy_class(InstanceKlass* caller_ik,
                                                    InstanceKlass* lambda_ik,
                                                    Symbol* invoked_name,
                                                    Symbol* invoked_type,
                                                    Symbol* method_type,
                                                    Method* member_method,
                                                    Symbol* instantiated_method_type
```
We use more parameters than `add_lambda_proxy_class` - I may be wrong but it seems like `caller` is skipped. Is this because we want the lambda names to be the same regardless of who created the metafactory?

Reply:
Caller is actually `targetClass` in `InnerClassLambdaMetafactory`. More precisely, it is a capturing class of a lambda which is already included as the name is generated as `CapturingClass$$Lambda$hashCodeForLambdaClass`. It can be included in the hashCode of course, if you think it is the better way.

Reply:
That sounds good, I think there's no need to include it in the hash in that case.